rclpy/rclpy/context.py

```diff
@@ -46,17 +46,34 @@ def __init__(self):
     def handle(self):
         return self._handle
 
-    def init(self, args: Optional[List[str]] = None, *, initialize_logging: bool = True):
+    def init(self,
+             args: Optional[List[str]] = None,
+             *,
+             initialize_logging: bool = True,
+             domain_id: Optional[int] = None):
         """
         Initialize ROS communications for a given context.
 
         :param args: List of command line arguments.
         """
         # imported locally to avoid loading extensions on module import
         from rclpy.impl.implementation_singleton import rclpy_implementation
+
         global g_logging_ref_count
         with self._handle as capsule, self._lock:
-            rclpy_implementation.rclpy_init(args if args is not None else sys.argv, capsule)
+            if domain_id is not None:
+                if domain_id < 0:
+                    raise RuntimeError(
+                        'Domain id ({}) should not be lower than zero.'
+                        .format(domain_id))
+                rclpy_implementation.rclpy_init(
+                    args if args is not None else sys.argv,
+                    capsule,
+                    domain_id)
+            else:
+                rclpy_implementation.rclpy_init(
+                    args if args is not None else sys.argv,
+                    capsule)
```

Comment:
I think that the logic can be simplified a bit:
```suggestion
            if domain_id is not None and domain_id < 0:
                raise RuntimeError(
                    'Domain id ({}) should not be lower than zero.'
                    .format(domain_id))
            rclpy_implementation.rclpy_init(
                args if args is not None else sys.argv,
                capsule,
                domain_id)
```

Reply:
@ivanpauno 

I tried this way. But while `domain_id` is None, code reports error as below.  
python thinks there is third argument. 

```shell
2: self = <rclpy.context.Context object at 0x7f7a2060eeb0>, args = None
2: 
2:     def init(self,
2:              args: Optional[List[str]] = None,
2:              *,
2:              initialize_logging: bool = True,
2:              domain_id: Optional[int] = None):
2:         """
2:         Initialize ROS communications for a given context.
2:     
2:         :param args: List of command line arguments.
2:         """
2:         # imported locally to avoid loading extensions on module import
2:         from rclpy.impl.implementation_singleton import rclpy_implementation
2:     
2:         global g_logging_ref_count
2:         with self._handle as capsule, self._lock:
2:             if domain_id is not None and domain_id < 0:
2:                 raise RuntimeError(
2:                     'Domain id ({}) should not be lower than zero.'
2:                     .format(domain_id))
2: >           rclpy_implementation.rclpy_init(
2:                 args if args is not None else sys.argv,
2:                 capsule,
2:                 domain_id)
2: E           TypeError: argument 3 must be int, not None
```

Reply:
After google, I find below solution.  

https://stackoverflow.com/questions/42333354/how-to-permit-none-or-a-specific-type-as-an-argument-to-a-python-c-extension-fun.  

Do you want to implement like above way?